### PR TITLE
Use templates with k_find_bound_blocks

### PR DESCRIPTION
**Pull request by badisa**
* 3x faster in float32, verified to produce same ixns as float64

The following is the profile, on a V100, of the new/old nb_list tests. Looking at the current `k_find_block_bounds` profile, it has the same performance as the templated double implementation which is to be expected. The float implementation is 3x faster on the T4, and 1.1X faster on the V100.
```
OLD
==14147== Profiling application: /home/ubuntu/.conda/envs/timemachine/bin/python /home/ubuntu/.conda/envs/timemachine/bin/pytest tests/test_nblist.py
==14147== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   54.52%  844.51us        12  70.375us  5.8560us  142.78us  void k_find_blocks_with_ixns<double>(int, double const *, double const *, double const *, double const *, unsigned int*, int*, unsigned int*, unsigned int*, double)
                   14.67%  227.17us        30  7.5720us  1.5040us  38.528us  [CUDA memcpy HtoD]
                   12.94%  200.41us        42  4.7710us  1.5990us  40.704us  [CUDA memcpy DtoH]
                   10.77%  166.88us        15  11.125us  10.304us  12.032us  k_find_block_bounds(int, int, int, double const *, double const *, double*, double*)
                    4.13%  63.904us        42  1.5210us  1.4400us  1.6960us  [CUDA memset]
                    2.98%  46.144us        12  3.8450us  3.5520us  4.2240us  k_compact_trim_atoms(int, int, unsigned int*, unsigned int*, int*, unsigned int*)

CHANGES
 GPU activities:   31.04%  845.08us        12  70.423us  5.8560us  142.98us  void k_find_blocks_with_ixns<double>(int, double const *, double const *, double const *, double const *, unsigned int*, int*, unsigned int*, unsigned int*, double)
                   23.34%  635.39us        12  52.948us  5.3760us  113.41us  void k_find_blocks_with_ixns<float>(int, double const *, double const *, double const *, double const *, unsigned int*, int*, unsigned int*, unsigned int*, double)
                   15.78%  429.66us        54  7.9560us  1.5360us  38.112us  [CUDA memcpy HtoD]
                   14.37%  391.26us        78  5.0160us  1.6000us  40.704us  [CUDA memcpy DtoH]
                    6.22%  169.28us        15  11.285us  10.432us  12.256us  void k_find_block_bounds<double>(int, int, int, double const *, double const *, double*, double*)
                    4.56%  124.10us        12  10.341us  9.5360us  11.360us  void k_find_block_bounds<float>(int, int, int, double const *, double const *, double*, double*)
                    3.36%  91.487us        24  3.8110us  3.4880us  4.2880us  k_compact_trim_atoms(int, int, unsigned int*, unsigned int*, int*, unsigned int*)
                    1.34%  36.351us        24  1.5140us  1.4720us  1.9200us  [CUDA memset]
```

### CODE REVIEW

File: tests/test_nblist.py [hunk 1]
```diff
@@ -69,68 +69,68 @@ def test_neighborlist():
 
     for size in [35, 64, 129, 1025, 1259, 2029]:
 
-        nblist = custom_ops.Neighborlist_f64(size)
-        water_coords = get_water_coords(3, sort=False)
+        for precision, nblist in ((np.float32, custom_ops.Neighborlist_f32(size)), (np.float64, custom_ops.Neighborlist_f64(size))):
+            water_coords = precision(get_water_coords(3, sort=False))
```
Comment (proteneer) [inline tests/test_nblist.py:73]: Is there a reason for casting the precision of the input coordinates? The input args always double

File: timemachine/cpp/src/neighborlist.cu [hunk 2]
```diff
@@ -218,10 +218,7 @@ void Neighborlist<RealType>::compute_block_bounds_device(
     const int tpb = 32;
     const int B = (N+tpb-1)/tpb; // total number of blocks we need to process
 
-    gpuErrchk(cudaMemsetAsync(d_block_bounds_ctr_, 0, B*D*sizeof(*d_block_bounds_ctr_), stream));
```
Comment (proteneer) [inline timemachine/cpp/src/neighborlist.cu:221]: nice! I didn't realize this was no longer necessary. The kernel directly initializes everything!

**Review by proteneer — APPROVED** [review]
lgtm!

File: tests/test_nblist.py (hunk 1 above)
Comment (badisa) [reply tests/test_nblist.py:73]: Nope

**Pull request merged**